powermanga.6
.TH POWERMANGA 6 2014-08-17
.SH NAME
powermanga \- shoot\-'em\-up game
.SH SYNOPSIS
.B powermanga "[OPTIONS]"
.SH DESCRIPTION
Powermanga is a shoot'em up game; you will have to fight more than 60 different types of enemies.
.P
You start with a ship, and enemies show up throwing bombs at you. Shoot them, and get better weapons for your ship (otherwise, you won't go too far!)
.P
To get better weapons, you will have to catch green and red gems. Whenever you pick a green gem, the highlighted option at the right side of the screen goes up one unit.
.P
The options are, from the bottom to the top:
.TP
- More speed
.TP
- More energy to your ship
If you get out of energy, you lose your ship. Losing your ship means either going back to the ship you were using before you "upgraded" it, or dying if you had no previous ship! You can check your energy level in the little bar at the bottom (the bar at the right is yours; the one at the left is your enemy's, in case you're fighting a big one)
.TP
- Several options for different weapons
.TP
- The last option gets you a better ship
.P
If the current option is "more power to the leftside straight weapons", for example, you may select it and they will immediately take effect.
.P
If the current option is already the last, and you pick a green or red gem, you will go back to the first one (but you will get an extra side-weapon). Also, whenever you pick an option, the current option goes back to the first one.
.SH THE GEMS
During the game, you will sometimes see gems falling down when an enemy is destroyed. You may catch those gems:
.TP
- Green gems
As we said before, makes the current option go up one unit.
.TP
- Red gems
Similar to the green gem, but it increases 2 units instead of 1.
.TP
- Pink gems
Gives energy to your ship. If your energy is at the maximum level, this gem will give you a full blast that destroys all enemies currently on the screen.
.TP
- Blue gems
They multiply the score points for each destroyed enemy by 2. If you get the blue gem again, the score points are multiplied by 4; you can see the "2x" or "4x" blinking at the sidebar.
.TP
- Yellow gems
These give you protection balls. If you already have six protection balls and you get another yellow gem, you will get a special weapon!
.SH CHANGING SHIPS
When you get a better ship, you will be able to accumulate more energy, becoming more resistant to attacks. But whenever you "upgrade" your ship, a few things will happen:
.TP
- Protection
You will lose all your protection balls (let's say they're not "compatible" with your new ship).
.TP
- Weapons
You will lose one "degree" in your weapons (if you had used green gems to get the best possible weapon for shooting straight forward, for example, after upgrading your ship you will have to get one more green gem if you want your weapons at the maximum power.)
.TP
- Score
If your score was getting multiplied by 2 or 4, it won't be anymore (go get more blue gems!)
.SH KEYS
In the main menu, use the arrow keys and enter. To play the game, use the arrow keys to move, space for shooting and control to pick a selected option.
.P
Also:
.P
 CTRL+S           enable/disable the music
 CTRL+Q           finish the play current
 CTRL+A           about Powermanga
 ALT+F4           quit Powermanga
 P or Pause/Break enable/disable pause
 F or F11         switch between full screen and windowed mode
 PAGE UP          volume up
 PAGE DOWN        volume down
.SH OPTIONS
Valid command line options:
.TP
.B \-h, --help
prints help message and exit
.TP
.B \--version
prints version information and exit
.TP
.B \--window
windowed mode (full screen by default)
.TP
.B \--2x
scale2x effects which scale the image of 2x 
.TP
.B \--3x
scale3x effects which scale the image of 3x 
.TP
.B \--4x
scale4x effects which scale the image of 4x 
.TP
.B \--320
the game will run in a 320*200 window
.TP
.B \--fullscreen
the game will run in full screen 
.TP
.B \-v
verbose mode
.TP
.B \--verbose
verbose mode (with more messages)
.TP
.B \--nosound
disable sound
.TP
.B \--sound
enable sound
.TP
.B \--nosync
disable timer
.TP
.B \--easy
make the game easier by giving more bonuses
.TP
.B \--hard
make the game harder
.SH FILES
.B /usr/share/games/powermanga/graphics/
.br
.RS
Graphics files.
.RE
.B /usr/share/games/powermanga/data/
.br
.RS
Levels and data files.
.RE
.B /usr/share/games/powermanga/sound/
.br
.RS
Sound files.
.RE
.B /usr/share/games/powermanga/texts/
.br
.RS
Texts files.
.RE
.SH ENVIRONMENT
.TP
.B HOME
Used to search for graphics and sounds.
.SH AUTHOR
Powermanga - Copyright TLK Games. You are free to redistribute
and/or change it under the terms of the GNU General Public License. See the
.I /usr/share/doc/powermanga/copyright
for details.
.br
This manual page was written by Jeronimo Pellegrini <j_p@aleph0.info>, based on the previous one by Sam Hocevar <sam@zoy.org>, for the Debian GNU/Linux system (but may be used by others).


CHANGES
2015-01-10
Powermanga 0.93.1
- Fixes the speed of the stars after a new game that follows congratulation.
- Clears all gems after the final boss. Thanks to Josh Triplett for the report
  (Debian #764009)
- Disable insecure temporary file "/tmp/powermanga-log.txt". Thanks to 
  Josh Triplett for the report (Debian #764144)
- Translation of the man page in French.
- Fix file permissions in HTML directory. Thanks to  nemysisbsd for
  the report.
- Allows the joystick to be configured either from the command line or
  from the config file. Thanks to Maurerp for the patch.

2014-09-20
Powermanga 0.93
- Fix man page: The option to play the game in fullscreen mode
   must be --fullscreen not --full. (Patch Debian)
- Added keywords and a comment in German and French to the desktop file
  (Patch Debian)
- Fix configure.ac: Add custom CFLAGS (Patch Debian)
- Fix segmentation fault in the About menu (English version) 
- Fixes wrong joystick behaviour in display_sdl.c and allows the ship to move
  to the left side again. Thanks to Kalle Olavi Niemitalo for the report
  (Debian #561670)
- Allow compilation with Clang

2014-08-19
Powermanga 0.92
- Fix Debian bug #478213
- Fix configure.ac: replace AM_CONFIG_HEADER to AC_CONFIG_HEADERS
- Add  explicitly link with needed libm, fix Debian bug #632945
- Fix PNG_iTXt_SUPPORTED support
- Add italian language
- Fix negative coordinates explosions that caused a segmentation fault.

2012-08-26
Powermanga 0.91
- Update "configure.ac" file. 
- Fix compilation warnings (GCC 4.6.3 and Visual Studio)
- Fix alignment constraint (ARM and MIPS processors)
- Adds the ability to export all the game's graphics in PNG files
  using the command line.
- Build test (without sound support) with Visual Studio and run
  successfully on Windows Mobile (HTC Touch P3450) 
- Recognizes all the joysticks connected at startup
- Rewrites the file "music_game.zik" with Milkytracker.
  Now the library "SDL_mixer" can read and play this module.
- The application switches to pause when it loses focus.
- Minor bugfixes and improvements

2007-09-02
Powermanga 0.90
- the sounds were updated to replace non-free sounds
- joystick support was added
- scale2x, scale3x, and scale4x effects were added
- the order text was changed to support both plain text and keystrokes file
- the source code files were renamed
- the code was considerably cleaned up and rewritten
- comments and variables were translated to the English language
- all C++ code was removed and the remaining code was made to conform with C99
- the license was changed to the GPLv3
- resources files were renamed
- volume control was added
- GP2X and PSP patches were merged
- the high-score file format was changed

2006-03-27
Powermanga 0.80
- compile with GCC 4.0.3
- fix palette conversion in 16 bits RGB color mode (non-IA32 architecture)
- add config file
- add french language support
- various bugs fixes
- compile, run, and test the project on AMD64 (Ubuntu Dapper Drake) 
- compile, run, and test the project on Win32 (Patrice Duhamel)
- upload the source code on CVS (http://sourceforge.net/projects/powermanga/) 

2004-08-01:
Powermanga 0.79
- fix some minor bugs and memory leaks (Patrice Duhamel and Jozef Behran)
- duplicate set of cursor movement keys on the numeric keypad (modus)
- disable F key (full screen) if input player name

2003-07-14: 
Powermanga 0.78
- use automake/autoconf (by Samuel Hocevar)
- remplacing XFree routines by SDL routines
- game run in fullscreen

2003-07-01: (Major release)
Powermanga 0.77
- rewrite of the major part of the assembler code in langage C (Samuel Hocevar)
- port to Linux PowerPC (a fix for big endian problems)
- switching of SEAL routines to the SDL_mixer (musics and sounds)
- fix for bug in loading PCX files  (Samuel Hocevar)
- fix for bug of closing the game's window (with the close widget)

2003-04-07:
Powermanga 0.76 
 - bug of the 'uninstall' rules : 
   'rm -f -r $(EXEDIR)' instead 'rm -f -r $(EXEDIR)/$(PROG)'
   Thank to Torbjörn Andersson torbjorn.e.andersson@tietoenator.com

2003-03-29:
Powermanga 0.75 (Minor release)
- fix memory leak (courbe.cpp, gardiens.cpp and grille.cpp) and fix bug
  in linuxroutines.cpp by Samuel Hocevar <sam@zoy.org>
- fix bug in "playanim.cpp", write the AUTHORS file, and force by default
  in mode 640x480 pixels by Bruno Ethvignot <bruno@tlk.fr>

2002-08-20:
Powermanga 0.74 (Minor release)
- Update the man page, by Jeronimo Pellegrini <pellegrini@mpcnet.com.br>
- Update the scroll-text of the menu (given by Samuel Hocevar <sam@zoy.org>)
- Update Makefile (pacth given by Samuel Hocevar <sam@zoy.org>)

2002-08-17:
Powermanga 0.73 (Minor release)
- Fix a file loading bug. Don't find files in 
  "/usr/share/games/powermanga" (patch by Guillaume Cottenceau)
- Update Makefile

2002-08-11:
Powermanga 0.72 (Minor release)
- Ident the code with indent program and cleaning of the code 
- Fix thread music bug. Sometimes the sound did not work. 
  Because the thread would run in loop. This error prevented the program
  from leaving.
- Fix bug at the end of game. The game was blocked after the animation
  of the end. The player should press a key to continue.
- update Makefile (set optimizations options)
- Disable inline option in Deplacement_Tir_X() and Deplacement_Tir_Y()
  functions (patch by Guillaume Cottenceau, for gcc 3.1)
- Add --easy / --hard options (patch by Laurent Duperval)

2000-12-02:
Powermanga 0.71c (Minor release)
- Update Makefile
- Save score in /var/game/powermanga.hi
- Fixed files installing bugs 

2000-11-27:
Powermanga 0.71 (Initial release)

1999-08-17:
* start the Linux port


trunk/src/powermanga.c
/** 
 * @file powermanga.c
 * @brief The main loop of the game Powermanga 
 * @date 2015-09-09 
 * @author Jean-Michel Martin de Santero
 * @author Bruno Ethvignot
 */
/* 
 * copyright (c) 1998-2015 TLK Games all rights reserved
 * $Id: powermanga.c,v 1.32 2012/08/26 17:09:14 gurumeditation Exp $
 *
 * Powermanga is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Powermanga is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
#include "config.h"
#include "powermanga.h"
#include "tools.h"
#include "images.h"
#include "congratulations.h"
#include "curve_phase.h"
#include "config_file.h"
#include "display.h"
#include "electrical_shock.h"
#include "enemies.h"
#include "bonus.h"
#include "energy_gauge.h"
#include "explosions.h"
#include "shots.h"
#include "extra_gun.h"
#include "gfx_wrapper.h"
#include "grid_phase.h"
#include "guardians.h"
#include "log_recorder.h"
#include "menu.h"
#include "menu_sections.h"
#include "meteors_phase.h"
#include "movie.h"
#include "options_panel.h"
#include "satellite_protections.h"
#include "scrolltext.h"
#include "shockwave.h"
#include "sdl_mixer.h"
#include "spaceship.h"
#include "spaceship_weapons.h"
#include "starfield.h"
#include "texts.h"
#include "text_overlay.h"

Sint32 global_counter;
/** Pause mode is enable */
bool player_pause;
/** TRUE if "PAUSE" string is currently displayed */
bool is_pause_draw;
/** Display "GAME OVER" sprites string */
bool gameover_enable = TRUE;
#ifdef DEVELOPPEMENT
bool mouse_here = TRUE;
bool curve_editor_enable = FALSE;
#endif
/** Current image index of the TLK Games logo */
Sint32 tlk_logo_image_index;
/** TRUE if flip to the nex image */
bool tlk_logo_is_next_image;
/** X coordinate of the TLK Games logo */
Sint32 tlk_logo_xcoord;
/** Y coordinate of the TLK Games logo */
Sint32 tlk_logo_ycoord;
/** TRUE if the TLK Games logo is enabled */
bool tlk_logo_is_move;
bitmap logotlk[TLKLOGO_MAXOF_IMAGES];

/**
 * The main loop of game
 * @return TRUE if it completed successfully or FALSE otherwise
 */
bool
update_frame (void)
{
  spaceship_struct *ship = spaceship_get ();
#ifdef DEVELOPPEMENT
  /* 1 = phase grid editor enable (don't work) */
  static bool grid_editor = FALSE;
#endif
  /* global frame counter */
  global_counter++;

  /* play start and congratulations animations files
   * ("movie_congratulation.gca" and "movie_introduction.gca") */
  if (movie_playing_switch != MOVIE_NOT_PLAYED)
    {
      if (!movie_player ())
        {
          LOG_ERR ("movie_player() failed!");
          return FALSE;
        }
      else
        {
          return TRUE;
        }
    }

  display_clear_offscreen ();

#ifdef __EMSCRIPTEN__
  lock_surface_game ();
  lock_surface_options ();
  lock_surface_scores ();
#endif

  /* restores the level of energy of the player spaceship */
  spaceship_energy_restore ();

  /* phase grid and curve phase editor (don't work) */
#ifdef DEVELOPPEMENT
  if (grid_editor || curve_editor_enable)
    {
      if (grid_editor)
        {
          Grid_Edit ();
        }
      else
        {
          courbe_editeur ();
        }
    }
  else
#endif

    /* pause or main menu enable */
  if (!player_pause && menu_status == MENU_OFF &&
        menu_section == NO_SECTION_SELECTED)
    {
      /* 
       * handle the phases of the game 
       */
      /* phase 2: grids (enemy wave like Space Invaders) */
      grid_handle ();
      /* phase 1: curves (little skirmish) */
      curve_phase ();
      /* phase 3: meteor storm */
      meteors_handle ();
    }

  /* draw the starfield background */
  starfield_handle ();

  /* handle bonus: green, red, yellow, blue and purple gems */
  bonus_handle ();

  /* handle protection satellites and extra gun of the player spaceship  */
  if (!gameover_enable && menu_section == NO_SECTION_SELECTED)
    {
      /* orbital protection satellites gravitate around player's spaceship */
      satellites_handle ();
      /* extra gun positioned on the sides */
      guns_handle ();
    }

  /* handle enemies */
  if (!is_congratulations_enabled)
    {
      /* handling of all the possible types of enemies */
      enemies_handle ();
    }
  else
    {
      /* congratulations, end of the game */
      congratulations ();
    }

  /* spaceship temporary invincibility  */
  spaceship_invincibility ();

  /* handle the powerful electrical shocks */
  electrical_shock ();

  /* draw the player's spaceship */
  spaceship_draw ();

  /* handle explosions */
  explosions_handle ();

  /* handle shots */
  shots_handle ();

  /* wait until all enemies are dead before jumping on next phase */
  if (num_of_enemies == 0 && !player_pause && menu_status == MENU_OFF
      && menu_section == NO_SECTION_SELECTED)
    {
      /* end of a guardian phase? */
      if (!guardian_finished ())
        {
          LOG_ERR ("guardian_finished() failed!");
          return FALSE;
        }
      /* check if curve phase is finished */
      curve_finished ();
      /* check if grid phase is finished */
      grid_finished ();
      /* check if meteors phase is finished */
      if (!meteors_finished ())
        {
          LOG_ERR ("meteors_finished failed!");
          return FALSE;
        }
    }

  /* display pixel mouse pointer */
#ifdef DEVELOPPEMENT
  if (mouse_here)
    {
      put_pixel (game_offscreen, mouse_x, mouse_y, 5);
    }
#endif

  /* draw powerful circular shock wave propagated by the player spaceship */
  shockwave_draw ();

  /* animations of the options box on the right options panel */
  option_execution ();

  /* handle high score table, game over, about and order sections */
  menu_sections_run ();

  /* display "PAUSE" chars sprites */
  if (is_pause_draw)
    {
      text_pause_draw ();
    }

  /* player's spaceship come */
  if (spaceship_appears_count > 0)
    {
      spaceship_appears_count--;
      /* accelerate the speed of the starfield */
      starfield_speed += 0.028f;
      if (starfield_speed > 2.0)
        {
          starfield_speed = 2.0;
        }
      /* invincibility time */
      spaceship_set_invincibility (SPACESHIP_INVINCIBILITY_TIME / 3);
    }

  /* player's spaceship disappearing to the bottom of the screen. */
  if (spaceship_disappears && is_congratulations_enabled == 0)
    {
      /* decelerate the speed of the starfield */
      starfield_speed -= 0.02f;
      if (starfield_speed <= 0.0)
        {
          starfield_speed = 0.0;
          ship->y_speed = 0.0;
        }
      else
        {
          /* accelerate the speed of the spaceship */
          ship->y_speed += -0.15f;
        }
    }

  /* display number level */
  text_level_draw ();

  /* display scrolltext in the main menu */
  scrolltext_handle ();

  /* handle the main menu of Powermanga */
  menu_handle ();

  /* [F1] spaceship_appears / [F2] spaceship disappears */
#ifdef DEVELOPPEMENT
  if (keys_down[K_F1])
    {
      spaceship_disappears = 1;
    }
  if (keys_down[K_F2] && starfield_speed == 0.0)
    {
      spaceship_show ();
    }
#endif

  /* handle "TLK Games" sprite logo */
  if (tlk_logo_is_move)
    {
      draw_bitmap (&logotlk[tlk_logo_image_index], tlk_logo_xcoord,
                   tlk_logo_ycoord);
      if (tlk_logo_is_next_image)
        {
          tlk_logo_image_index++;
          if (tlk_logo_image_index >= TLKLOGO_MAXOF_IMAGES)
            {
              tlk_logo_image_index = 0;
            }
          tlk_logo_is_next_image = FALSE;
          tlk_logo_ycoord--;
        }
      else
        {
          tlk_logo_is_next_image = TRUE;
        }
      if (tlk_logo_ycoord <= 32)
        {
          tlk_logo_is_move = FALSE;
        }
    }

  /* display text overlay (about, cheats menu and variables) */
  text_overlay_draw ();

  /* handle the loss and the regression of the spaceship or cause game over */
  spaceship_downgrading ();

  /* handle spaceship's energy level */
  energy_gauge_spaceship_update ();

  /* handle guardian's energy level */
  energy_gauge_guardian_update ();

  /* draw player's score into the top panel */
  text_draw_score ();

#ifdef DEVELOPPEMENT
  if (keys_down[K_E] && keys_down[K_G])
    {
      /* [e]+[g] = enable the grid editor */
      grid_editor = 1;
    }
  if (keys_down[K_E] && keys_down[K_C])
    {
      /* [e]+[c] enable the curve editor */
      curve_editor_enable = 1;
    }
  /* [p]+[g] back to the game */
  if (keys_down[K_P] && keys_down[K_G])
    {
      grid_editor = 0;
      curve_editor_enable = 0;
    }
#endif

  /* control the spaceship movements */
  if (!player_pause && !gameover_enable && menu_status == MENU_OFF
      && menu_section == NO_SECTION_SELECTED)
    {
      spaceship_control_movements ();
    }

  /* [P] or [Pause] enable/disable pause */
  if (!keys_down[K_CTRL] && (keys_down[K_P] || keys_down[K_PAUSE]))
    {
      toggle_pause ();
    }
  keys_down[K_P] = FALSE;
  keys_down[K_PAUSE] = FALSE;   /* clear flag pause key */


  /* switch between full screen and windowed mode */
#ifdef POWERMANGA_SDL
  if ((keys_down[K_F] && !is_playername_input ()
       && menu_section != SECTION_ORDER) || keys_down[K_F11])
    {
      if (power_conf->fullscreen)
        {
          power_conf->fullscreen = FALSE;
        }
      else
        {
          power_conf->fullscreen = TRUE;
        }
      init_video_mode ();
    }
  keys_down[K_F] = FALSE;
  keys_down[K_F11] = FALSE;
#endif

  /* control the speed of the spaceship */
  spaceship_speed_control ();

  /* cheat code keys */
#ifdef UNDER_DEVELOPMENT
  special_keys ();
#endif

  /* handle weapon's player spaceship */
#ifdef DEVELOPPEMENT
  if (!grid_editor && !curve_editor_enable && !spaceship_is_dead
      && !player_pause && menu_status == MENU_OFF
      && menu_section == NO_SECTION_SELECTED)
#else
  if (!spaceship_is_dead && !player_pause && menu_status == MENU_OFF
      && menu_section == NO_SECTION_SELECTED)
#endif
    {
      spaceship_weapons ();
    }

#ifdef __EMSCRIPTEN__
  unlock_surface_scores ();
  unlock_surface_options ();
  unlock_surface_game ();
#endif

  return TRUE;
}

/**
 * Enable/disable pause
 **/
bool
toggle_pause (void)
{
  if (!spaceship_is_dead && spaceship_appears_count <= 0
      && !spaceship_disappears && menu_status == MENU_OFF
      && menu_section == NO_SECTION_SELECTED)
    {
      if (player_pause)
        {
          player_pause = FALSE;
        }
      else
        {
          player_pause = TRUE;
        }
      if (is_pause_draw)
        {
          is_pause_draw = FALSE;
        }
      else
        {
          is_pause_draw = TRUE;
        }
    }
  return player_pause;
}
